sdk/servicebus/azure-servicebus/azure/servicebus/_pyamqp/aio/_client_async.py

```diff
@@ -737,6 +736,8 @@ async def _client_ready_async(self):
             return False
         if self._link.get_state().value != 3:  # ATTACHED
             return False
+        # once the receiver client is ready/connection established, we set link_credit
+        self._link.link_credit = self._link_credit
```

Comment:
Should we be using `self._link.flow()` here?
Looking at the implementation for `Link`, `Sender` and `Receiver`, there's a couple of places where `self.link_credit` is used:
- As the default value when `flow()` is called, if an explicit `link_credit` value is not passed in.
- To set the current link credit on `attach_received`
- In the Sender only, during update pending delivery status.

So the only one that might cause problems here is the first one. Are we calling `flow()` explicitly in the SB or EH code anywhere? If so - do we pass in a link credit value, or rely on the fallback to the `self.link_credit` value provided at construction?

The reason I ask this is because it was my intention (eventually) that all these AMQP constructs (like, Link, Session, Connection) be immutable, and all these properties would no longer be settable. So I'm trying to understand if the existing methods are insufficient for what we need to do here.

Reply:
We do send a flow when [resetting link credit in `receive_messages`](https://github.com/Azure/azure-sdk-for-python/blob/a61ad1304e41bea51d3a648dcbf51cc79f8ac015/sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py#L412)] and explicity pass in link_credit to flow [[here](https://github.com/Azure/azure-sdk-for-python/blob/a61ad1304e41bea51d3a648dcbf51cc79f8ac015/sdk/servicebus/azure-servicebus/azure/servicebus/_transport/_pyamqp_transport.py#L734)]. So, this particular outgoing flow isn't a problem.

The only issue is when we call the receive iterator, it calls ReceiveClient.receive_messages_iter -->ReceiveClient.do_work --> which sends out a [flow frame without explicitly passing link_credit](https://github.com/Azure/azure-sdk-for-python/blob/a61ad1304e41bea51d3a648dcbf51cc79f8ac015/sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/client.py#L855). In the iterator case, the default flow frame would use the original link_credit value passed in of 0.

I think to have the iterator work and not touch the link.link_credit value after creation, we would need to explicitly pass in `link_credit=self._link_credit` to flow here: https://github.com/Azure/azure-sdk-for-python/blob/e0acd6fdb8b98200ab10bd401430acff14992f3b/sdk/servicebus/azure-servicebus/azure/servicebus/_pyamqp/client.py#L855. Then we can remove this line.

I can run tests with that change if that sounds reasonable?

Reply:
OOB discussion with Anna: this seems reasonable.